### PR TITLE
cleanup bound variable handling

### PR DESCRIPTION
**Pull request by lcnr**
each commit should be pretty self-contained and hopefully straightforward to review.

I've added 677ec23a8dbf8ff5f1c03ccebd46f8b85e5ec1fc so that we can stop returning the region map from `replace_bound_vars_with_fresh_vars` in the following commit.

r? @nikomatsakis or @jackh726

### CODE REVIEW

**Comment by jackh726** [comment]
r=me with clean perf

File: compiler/rustc_infer/src/infer/higher_ranked/mod.rs [hunk 1]
```diff
@@ -76,21 +83,25 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         // (i.e., if there are no placeholders).
         let next_universe = self.universe().next_universe();
 
+        let replaced_bound_var = Cell::new(false);
```
Comment (jackh726) [inline compiler/rustc_infer/src/infer/higher_ranked/mod.rs:86]: I wouldn't be surprised if this is the cause of the perf regression.

File: compiler/rustc_infer/src/infer/mod.rs [hunk 2]
```diff
@@ -1524,25 +1523,40 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
...
-            |br: ty::BoundRegion| self.next_region_var(LateBoundRegion(span, br.kind, lbrct));
-        let fld_t = |_| {
-            self.next_ty_var(TypeVariableOrigin {
-                kind: TypeVariableOriginKind::MiscVariable,
-                span,
+        if let Some(inner) = value.no_bound_vars() {
+            return inner;
+        }
+
+        let mut region_map = FxHashMap::default();
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/mod.rs:1534]: it looks like the perf regression is caused by us always creating 3 separate storages here and then having to drop them.

Went from `BTreeMap` -> `SsoHashMap` -> `FxHashMap` to figure out if these help. The more complex `Drop` impl (and greater size) of `SsoHashMap` probably made things worse. So hopefully the comparatively simple `HashMap` ends up being an improvement.

If not I think we should commit to tracking bound vars in binders and modify this to eagerly create a `Vec` of inference vars using `value.bound_vars()` and then using that. Doing so should allow us to get some perf improvements in other areas as well.

**Comment by lcnr** [comment]
perf is pretty stupid :shrug: whatever

@bors r=jackh726

**Comment by pnkfelix** [comment]
* mixed results, but the improvements far outweigh the regressions here.
* marking as triaged

@rustbot label: +perf-regression-triaged.

**Pull request merged**